### ISSUE TITLE
`unncessary_cast` is not triggered by `ptr::cast`

### ISSUE THREAD

**Issue by Rua** [rust-lang/rust-clippy#15624]
### Summary

When casting pointers to the same type using `ptr::cast` rather than the `as` operator, the `unnecessary_cast` lint does not trigger.

### Lint Name

`unnecessary_cast`

### Reproducer

I tried this code:

```rust
fn main() {
    let foo: *mut u32 = std::ptr::null_mut();
    let bar: *mut u32 = foo.cast();
    let baz = foo.cast::<u32>();
}
```

I expected to see this happen:
Both the `bar` (with type inference) and `baz` (with explicit type) lines should trigger the lint.

Instead, this happened:
The lint was not triggered.

### Version

```text
rustc 1.89.0 (29483883e 2025-08-04)
binary: rustc
commit-hash: 29483883eed69d5fb4db01964cdf2af4d86e9cb2
commit-date: 2025-08-04
host: x86_64-unknown-linux-gnu
release: 1.89.0
LLVM version: 20.1.7
```

**Comment by KisaragiEffective** [comment]
It looks like `unnecessary_cast` only looks expressions form of `expr as Ty` where `let expr: Ty`.
It might be better to expand this concept to the `cast` method.

**Comment by profetia** [comment]
I think this is an enhancement rahter than a FN

@rustbot label -C-bug +C-enhancement
